### PR TITLE
[dpdk] Fix: update operands of count statement

### PR DESCRIPTION
**Pull request by kamleshbhalui**

### CODE REVIEW

**Comment by usha1830** [comment]
Can we also check if there are any other missing cases?

**Comment by kamleshbhalui** [comment]
done

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 1]
```diff
@@ -371,6 +377,119 @@ CopyPropagationAndElimination::copyPropAndDeadCodeElim(
...
+            instr.push_back(new IR::DpdkRecircidStatement(replaceIfCopy(recd->pass, false)));
+       } else if (auto rarm = stmt->to<IR::DpdkRearmStatement>()) {
+            instr.push_back(new IR::DpdkRearmStatement(replaceIfCopy(rarm->timeout)));
+       } else if (auto csum = stmt->to<IR::DpdkChecksumAddStatement>()) {
+            instr.push_back(new IR::DpdkChecksumAddStatement(csum->csum, csum->intermediate_value,
+                    replaceIfCopy(csum->field)));
+       } else if (auto csum = stmt->to<IR::DpdkChecksumSubStatement>()) {
+            instr.push_back(new IR::DpdkChecksumSubStatement(csum->csum, csum->intermediate_value,
+                    replaceIfCopy(csum->field)));
+       } else if (auto hash = stmt->to<IR::DpdkGetHashStatement>()) {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.cpp:466]: Lets not remove moves inserted for Hash. Hash parameters are moved to metadata if they are not consecutive.

**Review by usha1830 — COMMENTED** [review]
Please check the cases where operands are explicitly moved to metadata because of dpdk instruction limitations. Such cases should not be removed

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.cpp:466]: ok

**Review by usha1830 — APPROVED** [review]
Looks fine to me

**Review by psivanup — APPROVED** [review]
Looks fine.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 2]
```diff
@@ -355,22 +355,100 @@ CopyPropagationAndElimination::copyPropAndDeadCodeElim(
         } else if (auto jc = stmt->to<IR::DpdkJmpLessStatement>()) {
             instr.push_back(new IR::DpdkJmpLessStatement(jc->label,
                         replaceIfCopy(jc->src1, false), replaceIfCopy(jc->src2)));
+        } else if (auto jc = stmt->to<IR::DpdkJmpLessOrEqualStatement>()) {
```
Comment (syogender) [inline backends/dpdk/dpdkAsmOpt.cpp:358]: Just curious. for safe programming, shouldn't we be using is<TYPE> before converting to to<TYPE>?

**Review by syogender — COMMENTED** [review]
we don't add testcases ?

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 2 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.cpp:358]: I think `is` is for when we do not really need the casting result but just the check., But when we need the casting result doing `is` first is unnecessary overhead.
Also this kind of style is prevalent in p4c source everywhere i.e. https://github.com/p4lang/p4c/blob/main/frontends/p4/typeMap.cpp#L23

Comment (usha1830) [reply backends/dpdk/dpdkAsmOpt.cpp:358]: The to<>() result is assigned to var, which is checked in if condition. This is equivalent to 
if (is<>())
  auto var = to<>()

**Comment by kamleshbhalui** [comment]
> we don't add testcases ?

we do add it, but for this we verified on a p4 source which I thought is proprietary, I will ask Usha if it can be shared in open source directly or modifying a little.

**Pull request closed without merging**